src/Shared/IInternable.cs

```diff
@@ -0,0 +1,346 @@
+        /// </summary>
+        public string ExpensiveConvertToString()
+        {
+            // PERF NOTE: This will be an allocation hot-spot because the StringBuilder is finally determined to
+            // not be internable. There is still only one conversion of StringBuilder into string it has just
+            // moved into this single spot.
+            return _target.ToString();
+        }
+
+        /// <summary>
+        /// Compare target to string. Assumes string is of equal or smaller length than target.
+        /// </summary>
+        public bool StartsWithStringByOrdinalComparison(string other)
+        {
+#if DEBUG
+            ErrorUtilities.VerifyThrow(other.Length <= _target.Length, "should be at most as long as target");
+#endif
+            int length = other.Length;
+
+            // Backwards because the end of the string is (by observation of Australian Government build) more likely to be different earlier in the loop.
+            // For example, C:\project1, C:\project2
+            for (int i = length - 1; i >= 0; --i)
+            {
+                if (_target[i] != other[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Don't use this function. Use ExpensiveConvertToString
+        /// </summary>
+        public override string ToString() => throw new InvalidOperationException();
+    }
+
+    /// <summary>
+    /// A wrapper over char[].
+    /// </summary>
+    internal struct CharArrayInternTarget : IInternable
+    {
+        /// <summary>
+        /// Start index for the string
+        /// </summary>
+        private readonly int _startIndex;
+
+        /// <summary>
+        /// The held array
+        /// </summary>
+        private readonly char[] _target;
+
+        /// <summary>
+        /// Pointless comment about constructor.
+        /// </summary>
+        internal CharArrayInternTarget(char[] target, int count)
+            : this(target, 0, count)
+        {
+        }
+
+        /// <summary>
+        /// Pointless comment about constructor.
+        /// </summary>
+        internal CharArrayInternTarget(char[] target, int startIndex, int count)
+        {
+#if DEBUG
+            if (startIndex + count > target.Length)
+            {
+                ErrorUtilities.ThrowInternalError("wrong length");
+            }
+#endif
+            _target = target;
+            _startIndex = startIndex;
+            Length = count;
+        }
+
+        /// <summary>
+        /// The length of the target.
+        /// </summary>
+        public int Length { get; }
+
+        /// <summary>
+        /// Indexer into the target. Presumed to be fast.
+        /// </summary>
+        public char this[int index]
+        {
+            get
+            {
+                if (index > _startIndex + Length - 1 || index < 0)
```

Comment:
Suggest removing this check. It's not adding any value and just makes the codepath longer.

Reply:
and SubstringInternTarget  doesn't have it

Reply:
Great point, removed!